### DISCUSSION TITLE
question: JDBC connection (from DataGrip

### DISCUSSION THREAD

**Discussion by geoHeil** [duckdb/duckdb#3322] [General]
How can I create a READONLY JDBC connection from a SQL tool like DataGrip?

From https://youtrack.jetbrains.com/issue/DBE-15099 I have downloaded and registered the JDBC driver as described there.
https://search.maven.org/artifact/org.duckdb/duckdb_jdbc lists some of the settings:

```
org.duckdb.DuckDBDriver

jdbc:duckdb:/tmp/my_database
Properties ro_prop = new Properties();
ro_prop.setProperty("duckdb.read_only", "true");
Connection conn_ro = DriverManager.getConnection("jdbc:duckdb:/tmp/my_database", ro_prop);
```

but I cannot specify these properties in DataGrip - or I would at least not be aware of them. How can I specify the properties in Datagrip directly? Perhaps in the JDBC string?

**Comment by hannes** [comment]
Most JDBC clients allow setting the properties somewhere, indeed DataGrip also does:

<img width="841" alt="Screenshot 2022-03-29 at 20 49 59" src="https://user-images.githubusercontent.com/227792/160683806-bd1a62f4-a426-45df-9c31-2f5ad404ae37.png">

**Reply by geoHeil** [reply]
Thanks!

**Comment by geoHeil** [comment]
For 0.3.2 DataGrip is throwing always SQLExceptions that the setCatalog operation somehow fails. Do I need to change something in the configuration? The data can be displayed nicely though.

**Comment by Jens-H** [comment]
Looks like the setCatalog method is not conforming to the description in the Java docs which say: 

> If the driver does not support catalogs, it will silently ignore this request. 

So instead of throwing the SQLFeatureNotSupportedException the method call has to be ignored. Though, if this will solve your problem is another topic ...

**Reply by Jens-H** [reply]
The PR for this would be here: #3434

**Comment by geoHeil** [comment]
Interestingly I still get resource contention issues for my duckdb instance when datagrip is connected in read-only mode:

i.e. a python client tries to connect in writable (readonly=False) mode whilst datagrip is connected in read-only mode and the python client throws: 

Resource temporarily unavailable when calling the connect method.

IS this intended? I thought multiple threads can connect in read only_mode (whilst having a single writer) modifying the DB

**Reply by Mytherin** [reply]
As many threads can connect in read only or read-write mode as is desirable. However, only *one process* can connect in read-write mode, and when a process is connected in read-write mode it locks out all other processes, including read-only processes.
